lib/puppet/http/session.rb

```diff
@@ -11,29 +12,21 @@ def route_to(name, ssl_context: nil)
     cached = @resolved_services[name]
     return cached if cached
 
-    errors = []
-
     @resolvers.each do |resolver|
       Puppet.debug("Resolving service '#{name}' using #{resolver.class}")
-      resolver.resolve(self, name) do |service|
-        begin
-          service.connect(ssl_context: ssl_context)
-          @resolved_services[name] = service
-          Puppet.debug("Resolved service '#{name}' to #{service.url}")
-          return service
-        rescue Puppet::HTTP::ConnectionError => e
-          errors << e
-          Puppet.debug("Connection to #{service.url} failed, trying next route: #{e.message}")
-        end
+      service = resolver.resolve(self, name, ssl_context: ssl_context)
+      if service
+        @resolved_services[name] = service
+        Puppet.debug("Resolved service '#{name}' to #{service.url}")
+        return service
       end
     end
 
-    errors.each { |e| Puppet.log_exception(e) }
-
+    @resolution_exceptions.each { |e| Puppet.log_exception(e) }
     raise Puppet::HTTP::RouteError, "No more routes to #{name}"
```

Comment:
I mucked with the code to play with this a bit more:

I added the ServerList resolver to the CA service, and I removed the `raise` in the ServerList resolver so that we would fall back to the settings resolver. I also ensured both the SRV and ServerList resolvers return `nil` if we don't find anything that connects successfully (otherwise we get weird errors when it tries to resolve `service.url` on line 20, since `service` is being set to either the array of servers in `server_list` or an empty array for srv).  

```
be ./bin/puppet agent -t --server_list apple.example.com,google.example.com --use_srv_records --debug
```
```
Debug: Resolving service 'ca' using Puppet::HTTP::Resolver::ServerList
Debug: Creating new connection for https://apple.example.com:8140
Debug: Starting connection for https://apple.example.com:8140
Debug: Unable to connect to server from server_list setting: Failed to connect to https://apple.example.com:8140/status/v1/simple/master: Failed to open TCP connection to apple.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Debug: Creating new connection for https://google.example.com:8140
Debug: Starting connection for https://google.example.com:8140
Debug: Unable to connect to server from server_list setting: Failed to connect to https://google.example.com:8140/status/v1/simple/master: Failed to open TCP connection to google.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Debug: Resolving service 'ca' using Puppet::HTTP::Resolver::SRV
Debug: Searching for SRV records for domain: Home
Debug: Found 0 SRV records for: _x-puppet-ca._tcp.Home
Debug: Searching for SRV records for domain: Home
Debug: Found 0 SRV records for: _x-puppet._tcp.Home
Debug: Resolving service 'ca' using Puppet::HTTP::Resolver::Settings
Debug: Creating new connection for https://puppet:8140
Debug: Starting connection for https://puppet:8140
Debug: Connection to https://puppet:8140/puppet-ca/v1 failed, trying next route: Failed to connect to https://puppet:8140/puppet-ca/v1: execution expired
Error: Failed to connect to https://apple.example.com:8140/status/v1/simple/master: Failed to open TCP connection to apple.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Wrapped exception:
Failed to open TCP connection to apple.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Error: Failed to connect to https://google.example.com:8140/status/v1/simple/master: Failed to open TCP connection to google.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Wrapped exception:
Failed to open TCP connection to google.example.com:8140 (getaddrinfo: nodename nor servname provided, or not known)
Error: Failed to connect to https://puppet:8140/puppet-ca/v1: execution expired
Wrapped exception:
execution expired
Error: No more routes to ca
Error: Could not run: No more routes to ca
```